[Merged by Bors] - feat(LinearAlgebra/Multilinear/Basic): derivative of a multilinear map

smorel394 opened the pull request:
- Define the derivative `f.linearDeriv` of a multilinear map `f` at `x`, i.e. the first-order term of `f(x+h)-f(x+h')`.
- Define `MultilinearMap.domDomRestrict`: given a multilinear map indexed by a type `ι`, a function `P: ι → Prop` and a vector `z` define on `{a // ¬ P a}`, this is the multilinear map indexed by `{a // P a}`  sending `x` to the value of `f` at the vector that has coordinate at `a : ι` equal to `x a` if `P a` and `z a` if ` ¬ P a`.

- [x] depends on: #9137

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg commented:
I took the liberty to label this PR by topic area, and tweak title and description to better match the standard format.
(Github's defaults are not always ideal here.)

smorel394 commented:
Thanks !

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -252,6 +254,21 @@ def toLinearMap [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) : M₁ i →ₗ[R]
+derivative.-/
+noncomputable def linearDeriv [DecidableEq ι] [Fintype ι] (f : MultilinearMap R M₁ M₂)
+(x : (i : ι) → M₁ i) : ((i : ι) → M₁ i) →ₗ[R] M₂ :=
+Finset.sum Finset.univ (fun (i : ι) => (f.toLinearMap x i).comp (LinearMap.proj i))
```
```suggestion
 ∑ i : ι. (f.toLinearMap x i).comp (LinearMap.proj i)
```

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -252,6 +254,21 @@ def toLinearMap [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) : M₁ i →ₗ[R]
+/-- This introduces the "derivative" of a multilinear map, as a linear map from
+`(i : ι) → M₁` to `M₂`. For continuous multilinear maps, this will indeed be the
+derivative.-/
+noncomputable def linearDeriv [DecidableEq ι] [Fintype ι] (f : MultilinearMap R M₁ M₂)
```
Why is this noncomputable?

smorel394 replied:
Because Lean originally complained about it. But I tried removing the "noncomputable" and it's fine.

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -252,6 +254,21 @@ def toLinearMap [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) : M₁ i →ₗ[R]
+
+@[simp]
+lemma linearDeriv_apply [DecidableEq ι] [Fintype ι] (f : MultilinearMap R M₁ M₂)
+    (x y : (i : ι) → M₁ i) : f.linearDeriv x y
+    = Finset.sum Finset.univ (fun (i : ι) => f (Function.update x i (y i))) := by
```
It's usually best to put both sides of an `=` on the same line if you can, for readability:
```suggestion
    (x y : (i : ι) → M₁ i) :
    f.linearDeriv x y = ∑ i : ι, f (Function.update x i (y i)) := by
```
(and notation again!)

eric-wieser commented:
> I took the liberty to [...] tweak the title [...] to better match the standard format

A very minor point: we don't put a space before the first `(`.

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -1194,6 +1211,330 @@ theorem map_sub [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) (x y : M₁ i) :
+and expresses the result using this order isomorphism.-/
+lemma map_sub_finset [LinearOrder ι] (f : MultilinearMap R M₁ M₂) (a b v : (i : ι) → (M₁ i))
+    {s : Finset ι} (hs : Finset.card s = n) :
+    f (s.piecewise a v) - f (s.piecewise b v) = Finset.sum Finset.univ (fun (i : Fin n) =>
```
There's another sum notation opportunity hiding here

smorel394 replied:
Changed.

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -1194,6 +1211,330 @@ theorem map_sub [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) (x y : M₁ i) :
+    {s : Finset ι} (hs : Finset.card s = n) :
+    f (s.piecewise a v) - f (s.piecewise b v) = Finset.sum Finset.univ (fun (i : Fin n) =>
+    f (fun j => if h : j ∈ s then (if (s.orderIsoOfFin hs).symm ⟨j, h⟩ < i then a j
+    else if (s.orderIsoOfFin hs).symm ⟨j, h⟩ = i then a j - b j else b j) else v j)) := by
```
I think this lemma is till true if you replace `s.orderIsoOfFin hs` with `σ : Fin (s.card) ≃ s` (or perhaps keep `n` and use  `σ : Fin n ≃ s`) throughout, and will be a bit easier to read as a result.

smorel394 replied:
This was superseded by alreadydone's simplification.

alreadydone commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -1194,6 +1211,330 @@ theorem map_sub [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) (x y : M₁ i) :
+    intro s
+    simp only [Set.Finite.toFinset_setOf, Finset.mem_univ, forall_true_left, not_le,
+      Finset.compl_filter, not_lt, Finset.mem_filter, true_and]
+    rw [Nat.lt_succ_iff]
+  have heS : ∅ ∈ S := by rw [hS]; simp only [Finset.card_empty, Nat.zero_le]
+  have hS' : ∀ (s : Finset ι), s ∈ S.erase ∅ ↔ s.card = 1 := by
+    intro s
+    rw [Finset.mem_erase, hS, Nat.le_one_iff_eq_zero_or_eq_one, Finset.card_eq_zero]
+    aesop
+  rw [← (Finset.sum_erase_add _ _ heS), Finset.piecewise_empty, Finset.piecewise_empty,
+    sub_self, add_zero]
+  set I : (s : Finset ι) → (s ∈ S.erase ∅) → ι :=
+    fun s hs => by rw [hS' s, Finset.card_eq_one] at hs
+                   exact Classical.choose hs
+  have hI : ∀ (s : Finset ι) (hs : s ∈ S.erase ∅), I s hs ∈ Finset.univ :=
+    fun _ _ => Finset.mem_univ _
+  have heq : ∀ (s : Finset ι) (hs : s ∈ S.erase ∅), f (s.piecewise h x) - f (s.piecewise h' x) =
+      f (Function.update x (I s hs) ((h - h') (I s hs))) := by
+    intro s hs
+    rw [hS', Finset.card_eq_one] at hs
+    conv => lhs
+            rw [Classical.choose_spec hs, Finset.piecewise_singleton,
+              Finset.piecewise_singleton, ← MultilinearMap.map_sub]
+  set J : (i : ι) → (i ∈ Finset.univ) → Finset ι := fun i _ => {i}
+  have hJ : ∀ (i : ι) (hi : i ∈ Finset.univ), J i hi ∈ S.erase ∅ :=
+    fun _ _ => by rw [hS']; exact Finset.card_singleton _
+  have hJI : ∀ (s : Finset ι) (hs : s ∈ S.erase ∅), J (I s hs) (hI s hs) = s := by
+    intro s hs
+    simp only [Set.coe_setOf, Set.mem_setOf_eq]
+    rw [hS', Finset.card_eq_one] at hs
+    exact Eq.symm (Classical.choose_spec hs)
+  have hIJ : ∀ (i : ι) (hi : i ∈ Finset.univ), I (J i hi) (hJ i hi) = i := by
+    intro i hi; apply Eq.symm; rw [← Finset.mem_singleton]
+    have hs := hJ i hi
+    rw [hS', Finset.card_eq_one] at hs
+    change i ∈ {Classical.choose hs}
+    rw [← (Classical.choose_spec hs)]
+    simp only [Finset.mem_singleton]
+  rw [Finset.sum_bij' I hI heq J hJ hJI hIJ (g := fun i => f (Function.update x i ((h - h') i))),
+    sub_self]
```
I changed the statement a bit and golfed it with two lemmas; I'll PR the first soon.
```
open Finset in
theorem _root_.Finset.powersetCard_one {α} [DecidableEq α] {s : Finset α} :
    s.powersetCard 1 = s.map ⟨_, Finset.singleton_injective⟩ := by
  ext t
  rw [mem_powersetCard, card_eq_one, mem_map, ← exists_and_left, exists_congr]
  intro a
  rw [eq_comm, ← s.singleton_subset_iff, and_congr_left_eq]; rfl
  rintro rfl; rfl

open Finset in
lemma map_add_sub_map [DecidableEq ι] [Fintype ι] (x h : (i : ι) → M₁ i) : 
    f (x + h) = f x + f.linearDeriv x h +
      ∑ s in univ.powerset.filter (2 ≤ ·.card), f (s.piecewise h x) := by
  rw [add_comm, map_add_univ, ← Finset.powerset_univ,
      ← sum_filter_add_sum_filter_not _ (2 ≤ ·.card)]
  simp_rw [not_le, Nat.lt_succ, le_iff_lt_or_eq (b := 1), Nat.lt_one_iff, filter_or,
    ← powersetCard_eq_filter, sum_union (univ.pairwise_disjoint_powersetCard zero_ne_one),
    powersetCard_zero, powersetCard_one, sum_singleton, Finset.piecewise_empty, sum_map,
    Function.Embedding.coeFn_mk, Finset.piecewise_singleton, linearDeriv_apply, add_comm]

open Finset in
lemma map_add_sub_map_add_sub_linearDeriv [DecidableEq ι] [Fintype ι] (x h h' : (i : ι) → M₁ i) :
    f (x + h) - f (x + h') - f.linearDeriv x (h - h') =
    ∑ s in univ.powerset.filter (2 ≤ ·.card), (f (s.piecewise h x) - f (s.piecewise h' x)) := by
  simp_rw [map_add_sub_map, add_assoc, add_sub_add_comm, sub_self, zero_add, ← LinearMap.map_sub,
    add_sub_cancel', sum_sub_distrib]
```

smorel394 replied:
Made the change and pushed, thanks !

alreadydone commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -1194,6 +1211,330 @@ theorem map_sub [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) (x y : M₁ i) :
   rw [sub_eq_add_neg, sub_eq_add_neg, MultilinearMap.map_add, map_neg]
 #align multilinear_map.map_sub MultilinearMap.map_sub
 
+/-- This calculates the differences between the values of a multilinear map at
+two arguments that differ on a finset `s` of `ι` of cardinality `n`. It requires a
+linear order on `ι` in order to choose an order isomorphism between `s` and `Fin n`
+and expresses the result using this order isomorphism.-/
+lemma map_sub_finset [LinearOrder ι] (f : MultilinearMap R M₁ M₂) (a b v : (i : ι) → (M₁ i))
+    {s : Finset ι} (hs : Finset.card s = n) :
+    f (s.piecewise a v) - f (s.piecewise b v) = Finset.sum Finset.univ (fun (i : Fin n) =>
+    f (fun j => if h : j ∈ s then (if (s.orderIsoOfFin hs).symm ⟨j, h⟩ < i then a j
+    else if (s.orderIsoOfFin hs).symm ⟨j, h⟩ = i then a j - b j else b j) else v j)) := by
```
I don't understand why this is not stated with simply `f a - f b` as the LHS, without mentioning a Finset `s` at all. When `a` only differ from `b` on a Finset `s`, all summands outside of `s` would vanish by linearity, and you get a sum over `s.card` terms, and I think you can still get estimates as easily.

alreadydone replied:
Okay, I now see that induction on the Finset `s` is probably the easiest way to prove such statements, and we also don't want to restrict ourselves to finite `ι`. I have been able to give a short proof to a very similar result that doesn't require an Equiv with `Fin n`. The version with two `piecewise` can be obtained by combining two copies of `map_sub_map` below.
```
lemma map_update [DecidableEq ι] (x : (i : ι) → M₁ i) (i : ι) (v : M₁ i)  :
    f (update x i v) = f x - f (update x i (x i - v)) := by
  rw [map_sub, update_eq_self, sub_sub_cancel]

open Finset in
lemma map_sub_map [LinearOrder ι] (a b : (i : ι) → M₁ i) (s : Finset ι) :
    f a - f (s.piecewise b a) =
    ∑ i in s, f (fun j ↦ if j ∈ s → j < i then a j else if i = j then a j - b j else b j) := by
  refine s.induction_on_min ?_ fun k s hk ih ↦ ?_
  · rw [Finset.piecewise_empty, sum_empty, sub_self]
  rw [Finset.piecewise_insert, map_update, ← sub_add, ih,
      add_comm, sum_insert (lt_irrefl _ <| hk k ·)]
  simp_rw [s.mem_insert]
  congr 1
  · congr; ext i; split_ifs with h₁ h₂
    · rw [update_noteq, Finset.piecewise_eq_of_not_mem]
      · exact fun h ↦ (hk i h).not_lt (h₁ <| .inr h)
      · exact fun h ↦ (h₁ <| .inl h).ne h
    · cases h₂
      rw [update_same, s.piecewise_eq_of_not_mem _ _ (lt_irrefl _ <| hk k ·)]
    · push_neg at h₁
      rw [update_noteq (Ne.symm h₂), s.piecewise_eq_of_mem _ _ (h₁.1.resolve_left <| Ne.symm h₂)]
  · apply sum_congr rfl; intro i hi; congr; ext j; congr 1; apply propext
    simp_rw [imp_iff_not_or, not_or]; apply or_congr_left'
    intro h; rw [and_iff_right]; rintro rfl; exact h (hk i hi)
```

alreadydone replied:
Sorry, not by combining two copies, but by substitution (piecewise twice is the same as piecewise once). Updated my code above.

smorel394 replied:
Thanks ! Let me test it with the rest of the project (this is only a technical lemma used in a larger proof that continuous multilinear maps are differentiable, so I need it to work with the rest of the proof), and I'll get back to you.

smorel394 replied:
Yup it works, it actually simplifies the proof of the later results. :-) I'm rewriting the file to integrate your changes, thanks !

smorel394 replied:
Made the changes and pushed.

alreadydone commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -252,6 +254,21 @@ def toLinearMap [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) : M₁ i →ₗ[R]
+  ∑ i : ι, (f.toLinearMap x i).comp (LinearMap.proj i)
+
+@[simp]
+lemma linearDeriv_apply [DecidableEq ι] [Fintype ι] (f : MultilinearMap R M₁ M₂)
+    (x y : (i : ι) → M₁ i) :
+    f.linearDeriv x y = ∑ i, f (x.update i (y i)) := by
```
Sorry, this doesn't actually work, probably because `x` isn't a non-dependent function. Let me change it back.
```suggestion
lemma linearDeriv_apply [DecidableEq ι] [Fintype ι] (f : MultilinearMap R M₁ M₂)
    (x y : (i : ι) → M₁ i) :
    f.linearDeriv x y = ∑ i, f (update x i (y i)) := by
```

smorel394 commented:
I pushed another file to my branch and it got added to the PR, even though that was not my intention (this other file needs some work before it can become a PR). Does anybody know how to reverse this ?

alreadydone commented:
I've discarded that file. Regarding what you attempt to do in the file: we have [AnalyticAt.contDiffAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FDeriv/Analytic.html#AnalyticAt.contDiffAt) and [AnalyticAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Analytic/Basic.html#AnalyticAt) is defined in terms of [ContinuousMultilinearMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Module/Multilinear/Basic.html#ContinuousMultilinearMap). So I guess it shouldn't be hard to get what you want through that route?

smorel394 commented:
> I've discarded that file. Regarding what you attempt to do in the file: we have [AnalyticAt.contDiffAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FDeriv/Analytic.html#AnalyticAt.contDiffAt) and [AnalyticAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Analytic/Basic.html#AnalyticAt) is defined in terms of [ContinuousMultilinearMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Module/Multilinear/Basic.html#ContinuousMultilinearMap). So I guess it shouldn't be hard to get what you want through that route?

It depends what you mean by "not hard"; mathematically it's very easy, but in Lean it might be a pain. Certainly if I had known that analytic functions were in mathlib, I would have tried to go that route, because this is the best way.

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -1194,6 +1211,81 @@ theorem map_sub [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) (x y : M₁ i) :
+linear order on `ι` in order to express the result.-/
+lemma map_piecewise_sub_map_piecewise [LinearOrder ι] (a b v : (i : ι) → M₁ i) (s : Finset ι) :
+    f (s.piecewise a v) - f (s.piecewise b v) = ∑ i in s, f
+      fun j ↦ if j ∈ s then if j < i then a j else if j = i then a j - b j else b j else v j := by
```
This is probably a good place to us `ltbyCases`

alreadydone replied:
Is it? You'd have to write `fun _ ↦ a j` etc. which doesn't look clean ...

smorel394 commented:
> > I've discarded that file. Regarding what you attempt to do in the file: we have [AnalyticAt.contDiffAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FDeriv/Analytic.html#AnalyticAt.contDiffAt) and [AnalyticAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Analytic/Basic.html#AnalyticAt) is defined in terms of [ContinuousMultilinearMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Algebra/Module/Multilinear/Basic.html#ContinuousMultilinearMap). So I guess it shouldn't be hard to get what you want through that route?
> 
> It depends what you mean by "not hard"; mathematically it's very easy, but in Lean it might be a pain. Certainly if I had known that analytic functions were in mathlib, I would have tried to go that route, because this is the best way. (And yes, I did ask about what I wanted to do on Zulip before starting.)

Okay, I know what to do, I hope it won't be too painful.

alreadydone commented:
I found [your Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Differentiability.20of.20bounded.20multilinear.20maps.20.3F) (cc @ocfnash) and I do recall seeing it before. However, I had no idea Analytic is defined in terms of multilinear maps, even though I did know it's in mathlib. I only became aware 2 weeks ago due to my involvement in the Hairer challenge, when I looked up the definition of Analytic when proving [evaluations of multivariate polynomials are analytic](https://github.com/leanprover-community/mathlib4/pull/9138/commits/a285b7d9f14cab1758e6b483b3c2ce3843aea59c#diff-bfe5d0518e7c120eede815999c537da5265deb6dc656d384628065d022d2969fR311-R317). Sorry for being unable to help earlier and for the extra work!

smorel394 commented:
> I found [your Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/Differentiability.20of.20bounded.20multilinear.20maps.20.3F) (cc @ocfnash) and I do recall seeing it before. However, I had no idea Analytic is defined in terms of multilinear maps, even though I did know it's in mathlib. I only became aware 2 weeks ago due to my involvement in the Hairer challenge, when I looked up the definition of Analytic when proving [evaluations of multivariate polynomials are analytic](https://github.com/leanprover-community/mathlib4/pull/9138/commits/a285b7d9f14cab1758e6b483b3c2ce3843aea59c#diff-bfe5d0518e7c120eede815999c537da5265deb6dc656d384628065d022d2969fR311-R317). Sorry for being unable to connect the dots earlier and for the extra work!

Eh, it's not your fault. So anyway, I proved that continuous multilinear maps are analytic, however, to deduce that they are differentiable, I need the target F to be complete ! At least that's a hypothesis in [AnalyticAt.contDiffAt](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FDeriv/Analytic.html#AnalyticAt.contDiffAt), I haven't checked whether it's possible to remove. 
But my old proof works without that extra hypothesis on F ! So I guess I should still PR it, and also PR the proof that continuous multilinear maps are analytic, now that it's written ?

alreadydone commented:
That's nice! I only just worked out the informal argument why a multilinear form f (of finitely many factors) is analytic; you need to compose it with the (linear) projections to obtain f(p1, ..., pn) and put that as the nth term in the power series. As for the completeness assumption, I think it's removable and [FormalMultilinearSeries.changeOrigin_eval](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Analytic/Basic.html#FormalMultilinearSeries.changeOrigin_eval) seems to be the key lemma to generalize. I'll try later if you don't beat me to it!

alreadydone commented:
I now realized that the derivative of an analytic function may not even exist if the space isn't complete, let alone continuously differentiable; the original series converging can't guarantee the derivative doesn't fall into a hole. So I think your result for continuous linear maps will be a valuable addition to mathlib! Maybe we could generalize it to analytic functions with finitely many nonzero terms ("polynomial functions"?). As a different project, the derivative's power series is in mathlib as [FormalMultilinearSeries.shift](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FormalMultilinearSeries.html#FormalMultilinearSeries.shift), but apparently no one has shown it evaluates to the derivative yet ...

alreadydone commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -252,6 +254,21 @@ def toLinearMap [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) : M₁ i →ₗ[R]
 #align multilinear_map.to_linear_map MultilinearMap.toLinearMap
 #align multilinear_map.to_linear_map_to_add_hom_apply MultilinearMap.toLinearMap_apply
 
+/-- This introduces the "derivative" of a multilinear map, as a linear map from
```
```suggestion
/-- The "derivative" of a multilinear map, as a linear map from
```

alreadydone commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -252,6 +254,21 @@ def toLinearMap [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) : M₁ i →ₗ[R]
+@[simp]
+lemma linearDeriv_apply [DecidableEq ι] [Fintype ι] (f : MultilinearMap R M₁ M₂)
+    (x y : (i : ι) → M₁ i) :
+    f.linearDeriv x y = ∑ i, f (update x i (y i)) := by
+  unfold linearDeriv
+  simp only [LinearMap.coeFn_sum, LinearMap.coe_comp, LinearMap.coe_proj, Finset.sum_apply,
+    Function.comp_apply, Function.eval, toLinearMap_apply]
```
either
```suggestion
    f.linearDeriv x y = ∑ i, f (update x i (y i)) := LinearMap.sum_apply _ _ _
```
or
```suggestion
    f.linearDeriv x y = ∑ i, f (Function.update x i (y i)) := by simp [linearDeriv]
```

alreadydone commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -1194,6 +1211,71 @@ theorem map_sub [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) (x y : M₁ i) :
+    intro h; rw [and_iff_right]; rintro rfl; exact h (hk i hi)
+
+/-- This calculates the differences between the values of a multilinear map at
+two arguments that differ on a finset `s` of `ι` of cardinality `n`. It requires a
```
```suggestion
two arguments that differ on a finset `s` of `ι`. It requires a
```

alreadydone commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -1194,6 +1211,71 @@ theorem map_sub [DecidableEq ι] (m : ∀ i, M₁ i) (i : ι) (x y : M₁ i) :
   rw [sub_eq_add_neg, sub_eq_add_neg, MultilinearMap.map_add, map_neg]
 #align multilinear_map.map_sub MultilinearMap.map_sub
 
+
```
```suggestion
```

alreadydone commented:
Maybe we can replace the CompleteSpace conditions by the necessary Summable conditions. Then they'll apply to both CompleteSpaces (where we can use [Summable.of_norm](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Normed/Group/InfiniteSum.html#Summable.of_norm)) and our situation ("polynomial functions"), where we have summability by virtue of being finite sums.

smorel394 commented:
Yes, the completeness is only really used to prove the convergence of infinite power series. So maybe the simplest would be to have some HasFiniteFPowerSeriesOnBall structure which would cover both polynomial and multilinear functions. This should be stable by taking derivatives, and so we would get both analyticity and differentiability results with any completeness condition. I won't get to it today as I have a pretty full schedule.
By the way, the "continuous multilinear functions are analytic" code is there: [](https://github.com/smorel394/ExteriorPowers/blob/master/Multilinear.lean)

smorel394 commented:
Ok, I added some more stuff to prepare for the proof that continuous multilinear maps are analytic: the definition of what will be the formal multilinear series at `x` and a proof that it does calculate `f (x + y)` (with a finite sum). Also, the comparison of the degree `1` term with the previously defined derivative. I kept the previous derivative, because it definition doesn't use a linear order on `ι`, while the definition of the formal multilinear series does.

Also, I kept running into merge problems that for some reason duplicated the new code, so I hope that those are resolved now. It seems to be okay. And I broke mathlib in my branch when trying to update it (again :-(((( ), so nothing compiles locally for me, but I did check the new code in another lean project, and again hopefully it's okay. Now I'll try to fix my local mathlib.

smorel394 commented:
Yeah, it's not okay, the thing didn't build. Why do I break stuff every time I update mathlib ? I only wanted the new lemma from PR #8867 9137. :-(

eric-wieser commented on lake-manifest.json:
```diff
@@ -4,7 +4,7 @@
    "type": "git",
    "subDir": null,
-   "rev": "9dd24a3493cceefa2bede383f21e4ef548990b68",
+   "rev": "fadefd5a532a774dc47d00f9fa0bce3dc47ba2e1",
```
This looks like a bad merge

eric-wieser commented:
(the base branches above are to work around a github quirk caused by a weird (but ultimately harmless) merge that you did)

smorel394 commented:
Okay, after last night's git drama, and after a lot of help from Eric, and after I accidentally pushed the whole of mathlib4 to a mathlib3 branch, it seems that I have finally been able to successfully update this. So I added some code to prepare for the proof that continuous multilinear maps are analytic: the definition of what will be the formal multilinear series at x, i.e. the higher derivatives, and a proof that it does calculate f (x + y) (with a finite sum). Also, the comparison of the degree 1 term with the previously defined derivative. I kept the previous derivative, because it definition doesn't use a linear order on ι, while the definition of the formal multilinear series does.

As the new code uses things like `MultilinearMap.domDomCongr`, I moved it down in the file so it would come after that, and put it in its own section called 'Derivative`.

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,204 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+is `f` evaluated at the vector with `i`th coordinate `(z i) i` if `i ∈ s` and `x i` otherwise.-/
+def toMultilinearMap_set [DecidableEq ι] (f : MultilinearMap R M₁ M₂) (x : (i : ι) → M₁ i)
+    (s : Set ι) [(i : ι) → Decidable (i ∈ s)] :
+    MultilinearMap R (fun (_ : s) => (i : ι) → M₁ i) M₂ where
```
Are you sure you don't want
```suggestion
    MultilinearMap R (fun (i : s) => M₁ i) M₂ where
```
which avoids the "diagonalization" of `z`, replacing `z ⟨i, h⟩ i` with `z ⟨i, h⟩` below?

smorel394 replied:
I eventually want a [FormalMultilinearSeries](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FormalMultilinearSeries.html#FormalMultilinearSeries) that sums to `f`, when the multilinear map `f` is also continuous. As `f` is a map from `(i : ι) → M₁` to `M₂`, the nth term of the FormalMultilinearSeries must be a multilinear map on `Fin n →((i : ι) → M₁)` (and then we evaluate it on the diagonal when we sum). This was the most direct way I found to do that.

I certainly do *not* want to restrict `f` to the components in `s`, and I do want the result to depend on `x`. I wonder if I might do what I want to do in another way, which would avoid the ugly definition that you point out and avoid the choice of the linear order. I need to write stuff down on paper.

smorel394 replied:
So I don't need to choose an order on `ι`, just an equivalence of each `s` of cardinality `n` with `Fin n`. Which choice is ugliest is up for debate. The mathematical point hiding behind that is that the nth derivative is not unique in positive characteristic. (In characteristic 0, I could just take the symmetrization of the thing I defined and divide by `n!`.)
As for the original question, yes, I could start by defining a multilinear map on `(i : s) → M₁ i` and build the nth derivative from that. I guess the auxiliary construction could having an independent interest, so let me try.

eric-wieser replied:
Yes, I think my claim is mainly "you could factor this through the definition I propose", and that the factored definition could be called `domDomRestrict`

smorel394 replied:
Working on it right now.

smorel394 replied:
Done. Now let's see what CI makes of this.

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,165 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+
+/-- Given a multilinear map `f` on `(i : ι) → M i`, an element `x` of `(i : ι) → M i` and s
+set `s` of `ι`, construct a multilinear map on `s → ((i : ι) → Mi)` whose value at `z`
+is `f` evaluated at the vector with `i`th coordinate `(z i) i` if `i ∈ s` and `x i` otherwise.-/
```
```suggestion
is `f` evaluated at the vector with `i`th coordinate `z i` if `i ∈ s` and `x i` otherwise.-/
```

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,165 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+/-- Given a multilinear map `f` on `(i : ι) → M i`, an element `x` of `(i : ι) → M i` and s
+set `s` of `ι`, construct a multilinear map on `s → ((i : ι) → Mi)` whose value at `z`
+is `f` evaluated at the vector with `i`th coordinate `(z i) i` if `i ∈ s` and `x i` otherwise.-/
+def domDomRestrict [DecidableEq ι] (f : MultilinearMap R M₁ M₂) (x : (i : ι) → M₁ i)
+    (s : Set ι) [(i : ι) → Decidable (i ∈ s)] :
+    MultilinearMap R (fun (i : s) => M₁ i) M₂ where
+  toFun z := f (fun i ↦ if h : i ∈ s then z ⟨i, h⟩ else x i)
```
How much messier do things become if you generalize to
```suggestion
def domDomRestrict [DecidableEq ι] (f : MultilinearMap R M₁ M₂) (x : (i : sᶜ) → M₁ i)
    (s : Set ι) [(i : ι) → Decidable (i ∈ s)] :
    MultilinearMap R (fun (i : s) => M₁ i) M₂ where
  toFun z := f (fun i ↦ if h : i ∈ s then z ⟨i, h⟩ else x ⟨i, h⟩)
```

smorel394 replied:
Not sure. I'm nervous because of things like [this](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Multilinear/Basic.html#MultilinearMap.toLinearMap), where you take `m` defined on all of `ι` even though you don't actually have to. (Mathematically it doesn't matter what we choose here.)

smorel394 replied:
Doesn't seem to change much. I can make the change, I'm pretty indifferent to it, just not sure what is more compatible with standard mathlib practice.

smorel394 replied:
Made the change.

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,156 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+def toFormalMultilinearSeries_fixedDegree [DecidableEq ι] [Fintype ι] [LinearOrder ι]
+    (f : MultilinearMap R M₁ M₂) (x : (i : ι) → M₁ i) (n : ℕ) :
+    MultilinearMap R (fun (_ : Fin n) => (i : ι) → M₁ i) M₂ :=
+  ∑ s : univ.powerset.filter (·.card = n),--{s : Finset ι | s.card = n},
```
```suggestion
  ∑ s : univ.powersetLen n,
```
seems more appropriate here

smorel394 replied:
Lean does not seem to know that definition, I get an error ("invalid field 'powersetLen', the environment does not contain 'Finset.powersetLen'"). I also cannot find `Finset.powersetLen` in mathlib.

eric-wieser replied:
Oh, I think I mean `powersetCard`

smorel394 replied:
I made the change, it does simplify things a bit.

jcommelin commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,151 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+/-- Given a multilinear map `f` on `(i : ι) → M i`, an element `x` of `(i : ι) → M i` and s
+set `s` of `ι`, construct a multilinear map on `s → ((i : ι) → Mi)` whose value at `z`
+is `f` evaluated at the vector with `i`th coordinate `z i` if `i ∈ s` and `x i` otherwise.-/
+
```
```suggestion
```

jcommelin reviewed (commented):
minor whitespace comment

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,151 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+equivalence of the two definitions.-/
+
+lemma domDomRestrict_aux [DecidableEq ι] (s : Set ι) [(i : ι) → Decidable (i ∈ s)]
+    [DecidableEq s] (x : (i : s.compl)→ M₁ i) (z : (i : s) → M₁ i) (i : s)
```
Can you use the `compl` notation here, as in my original suggestion?

smorel394 replied:
If you mean the notation `sᶜ`, no I can't. I get an error message 
```failed to synthesize instance
  HasCompl (Type uι)
```
and I don't know how to get rid of it.

eric-wieser replied:
```suggestion
    [DecidableEq s] (x : (i : ↥sᶜ)→ M₁ i) (z : (i : s) → M₁ i) (i : s)
```
should work, which tells lean you are casting from something that isn't a type.

smorel394 replied:
Yup it works, I changed it ! Thanks !

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,151 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+lemma domDomRestrict_aux [DecidableEq ι] (s : Set ι) [(i : ι) → Decidable (i ∈ s)]
+    [DecidableEq s] (x : (i : s.compl)→ M₁ i) (z : (i : s) → M₁ i) (i : s)
+    (c : M₁ i) : (fun j ↦ if h : j ∈ s then Function.update z i c ⟨j, h⟩ else x ⟨j, h⟩) =
+    Function.update (fun j => if h : j ∈ s then z ⟨j, h⟩ else x ⟨j, h⟩) i c := by
```
@YaelDillies, seems this function is appearing again; perhaps we really should have `Function.piecewise` or `Subtype.piecewise`?

YaelDillies replied:
Should that replace `Set.piecewise`?

eric-wieser replied:
I think we can keep `Set.piecewise` for the non-dependent version

YaelDillies replied:
That's quite confusing. `Set.piecewise` *is* dependent, but not on the set membership.

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,151 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+open Finset in
+/-- This is the nth term of a formal multilinear series corresponding to the multilinear map `f`.
+We use a linear order on ι to identify all finsets of `ι` of cardinality `n` to `Fin n`.-/
+def toFormalMultilinearSeries_fixedDegree [DecidableEq ι] [Fintype ι] [LinearOrder ι]
```
Defs are not allowed to contain `_`s in their names

smorel394 replied:
Changed it to `toFormalMultilinearSeriesFixedDegree`.

eric-wieser commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,150 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+of degree `1` in `f.linearDeriv`, which doesn't require a linear order on `ι`, and prove the
+equivalence of the two definitions.-/
+
+lemma domDomRestrict_aux [DecidableEq ι] (s : Set ι) [(i : ι) → Decidable (i ∈ s)]
+    [DecidableEq s] (x : (i : ↑sᶜ)→ M₁ i) (z : (i : s) → M₁ i) (i : s)
+    (c : M₁ i) : (fun j ↦ if h : j ∈ s then Function.update z i c ⟨j, h⟩ else x ⟨j, h⟩) =
+    Function.update (fun j => if h : j ∈ s then z ⟨j, h⟩ else x ⟨j, h⟩) i c := by
```
Possibly:
```suggestion
lemma domDomRestrict_aux [DecidableEq ι] (P : ι → Prop) [DecidablePred p]
    [DecidableEq s] (x : {a // P a} → M₁ i) (z : {a // ¬ P a} → M₁ i) (i : s)
    (c : M₁ i) : (fun j ↦ if h : P j then Function.update x i c ⟨j, h⟩ else z ⟨j, h⟩) =
    Function.update (fun j => if h : j ∈ s then x ⟨j, h⟩ else z ⟨j, h⟩) i c := by
```
(note I also swapped `x` and `z` to put them in the obvious order)

smorel394 replied:
Doesn't change much afaik, I made the change.
I also remove the bits about formal multilinear series to apply @alreadydone's suggestion to only define the formal multilinear series explicitly at the origin and then use `changeOrigin` to get it elsewhere. So now all these parts are in the file that proves that continuous multilinear maps are analytic (not yet PR'ed, as I still have to finish some calculations).

smorel394 commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,72 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
+an element `z` of `(i : {a // ¬ P a}) → M₁ i`, construct a multilinear map on
+`(i : {a // P a}) → M₁ i)` whose value at `x` is `f` evaluated at the vector with `i`th coordinate
+`x i` if `P i` and `z i` otherwise.-/
+def domDomRestrict [DecidableEq ι] (f : MultilinearMap R M₁ M₂) (P : ι → Prop) [DecidablePred P]
```
Done.

smorel394 commented on Mathlib/LinearAlgebra/Multilinear/Basic.lean:
```diff
@@ -760,8 +763,72 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
 
 end
 
+
+section
+
```
The section was useless, I removed it.

smorel394 commented:
bors r+

smorel394 commented:
bors r+

alreadydone commented:
[This branch](https://github.com/leanprover-community/mathlib4/compare/linearDeriv_fderiv?expand=1) is what I consider some unfinished work in this PR. I'm done for today but feel free to fill the sorries and submit another PR.

alreadydone commented:
> [This branch](https://github.com/leanprover-community/mathlib4/compare/linearDeriv_fderiv?expand=1) is what I consider some unfinished work from this PR. I'm done for today but feel free to fill the sorries and submit another PR.

This has been PR'd as #9846. Maybe reviewers here could take a look at this follow-up PR?
